lib/manageiq/environment.rb

```diff
@@ -121,6 +121,46 @@ def self.bundler_version
       File.read(gemfile).match(/gem\s+['"]bundler['"],\s+['"](.+?)['"]/)[1]
     end
 
+    def self.warn_if_branch_out_of_date
+      return if %w(production test).include?(ENV['RAILS_ENV'].to_s)
+
+      Dir.chdir(APP_ROOT) do
+        system!("git fetch")
+
+        commits =
+          if upstream_tracking_branch.empty?
+            commits_behind_upstream
+          else
+            commits_behind_tracked_branch
+          end
+
+        if commits > 0
+          warn "\n== Your branch is #{commits} commits out of date, please pull or fetch + merge."
+        end
+      end
+    end
+
+    def self.upstream_tracking_branch
+      `git for-each-ref --format='%(upstream:short)' $(git symbolic-ref -q HEAD)`.strip
+    end
+
+    def self.commits_behind_tracked_branch
+      # Show short status + branch information:
+      # ## master...upstream/master [behind 10]
+      # git 1.8.5+ shows upstream branch status, catch STDERR just in case
+      command = 'git status -sb 2>&1 | grep -oE "\[behind.+\]" | grep -oE "[0-9]+"'
+      `#{command}`.strip.to_i
+    end
+
+    def self.commits_behind_upstream
+      # Assume your remote is called upstream, returns early if you're special
```

Comment:
For @bdunne ^

Reply:
Maybe "Assume there is a remotea remote called upstream..."

Reply:
Wait, you're ok with upstream right?  It's only origin you despise, right?

Reply:
`upstream` makes sense.  To me `origin` means the place from which the code originated (which would be upstream).  I get confused when people call a remote `origin` and it points to a fork (not the place from which the code originated), so I prefer not to use it.

#NamingIsHard